### DISCUSSION TITLE
98. Next Key Lock은 언제 걸릴까요?

### DISCUSSION THREAD

**Discussion by jongmee** [jewelry-box/cs-interview#102] [General]
Next Key Lock은 언제 걸릴까요?

**Comment by linirini** [comment]
Next Key Lock은 record lock과 gap lock을 조합한 락이다.
쿼리 조건문에 해당하는 row에 락을 잡기 위해 구간을 스캔한 후 해당하는 인덱스 레코드에 대해 락을 획득한다. 추가적으로 레코드의 바로 이전 공간에 대해서도 락을 획득하여 gap lock의 역할도 수행한다. 즉, 인덱스순으로 처음 레코드를 발견했을 때 바로 이전 공간 및 조건 군간의 어떤 빈 공간에 대해서도 변경이 불가능하다.
** [참고자료](https://syongsyong2.tistory.com/117)
![image](https://github.com/user-attachments/assets/f78b46d4-1433-4616-aaa8-85567cf6a127)

InnoDB의 디폴트 격리 수준 REPEATABLE READ에서는 phantom read를 방지하기 위해 기본적으로 next-key lock을 이용하여 락을 획득한다. 하지만 gap lock을 비활성화 한 경우 next-key lock을 사용할 수 없다.

**Reply by eun-byeol** [reply]
넥스트 키락이 사용되는 예시나 사용 시점이 추가되면 좋을 것 같아요☺️

**Reply by linirini** [reply]
### 넥스트 키 락이 사용되는 시점
1. REPEATABLE READ 격리 수준에서 쿼리가 조건 범위를 스캔하는 경우
  - 트랜잭션 중 동일한 쿼리를 실행했을 때 결과가 변경되지 않아야 하는 REPEATABLE READ 수준에서는 넥스트 키 락이 자동으로 적용된다.
  - 예: WHERE 조건으로 특정 범위를 조회하거나 정렬하는 경우.
2. INSERT, UPDATE, DELETE와 같은 변경 연산 중
  - 특정 조건에 따라 레코드를 변경하거나 삭제할 때 다른 트랜잭션이 동일한 조건의 레코드에 접근하지 못하도록 락이 적용된다.
3. 명시적으로 락을 요청하는 경우
  - SELECT ... FOR UPDATE 또는 SELECT ... LOCK IN SHARE MODE를 사용해 명시적으로 레코드 락을 설정할 때 넥스트 키 락이 적용될 수 있다.
